fix[#3]: 에러 응답 개선

hunnypooh opened the pull request:
resolved #3 

## 작업 개요
- 에러 응답 제대로 동작하도록 수정.
- swagger 추가
- PR 템플릿 추가

## 작업 사항
**controller단에서 보내는 CustomException 를 잘 응답했는데 갑자기 안됬던 이유** 
-오류 원인 : authorization 헤더가 없을 경우 JwtRequestFilter 에서 CustomException을 날릴 경우 filter라서 GlobalExceptionHandler 에서 처리 불가능. 이 문제해결하려고 GlobalExceptionHandler 수정하다가 ErrorResponse의 @getter를 삭제함. 그래서 Auth.service에서 날리는 예외도 처리 못했음
- 해결 방법 : @getter 붙임

**jwt filter에서 던지는 CustomException 처리** 
- 원인 : 위에 적음.
- 해결 방법 : JwtRequestFilter 내부에서 try-catch로 CustomException 날리고 응답 보냄.


## 스크린샷
<img width="994" alt="image" src="https://github.com/user-attachments/assets/f1801877-6599-4fd6-8ead-f5f7f40960b9">

hunnypooh committed:
fix: CustomException 핸들링 오류 수정 [#3]

- 오류 원인 : authorization 헤더가 없을 경우 JwtRequestFilter 에서 CustomException을 날릴 경우 filter라서 GlobalExceptionHandler 에서 처리 불가능. 이 문제해결하려고 GlobalExceptionHandler 수정하다가 ErrorResponse의 @Getter를 삭제함. 그래서 Auth.service에서 날리는 예외도 처리 못했음
- 해결 방법 : 1) @Getter 붙임 2) JwtRequestFilter 내부에서 try-catch로 CustomException 날리고 응답 보냄.

- 모든 에러 코드, 메세지는 ErrorDetails 파일에서 확인 가능.
- 보내는 에러 응답 정리

상황) authorization 헤더가 없을 경우 == jwt 토큰이 null인 경우
응답) E003
    (WebSecurityConfig.java 에서 excludeUrls에 명시되지 않은 모든 요청은 authorization 헤더로 jwt가 만들어져야 정상 수행 가능)

상황) 로그인 했는데 다시 로그인 (토큰 생성 수행x)
응답) E004

상황) 로그아웃 했는데 다시 로그아웃 (토큰 삭제 수행x)
응답) E005

hunnypooh committed:
feat: swagger 추가 [#3]

http://localhost:8080/swagger-ui/index.html
<-url 입력시 스웨거 확인 가능

hunnypooh committed:
fix: 없어졌던 PR 템플릿 추가 [#3]

kimyu0218 commented on src/main/java/com/brick/demo/auth/jwt/JwtRequestFilter.java:
```diff
@@ -25,19 +31,23 @@ public class JwtRequestFilter extends OncePerRequestFilter {
+      }
+
+      chain.doFilter(request, response);
+    } catch (CustomException ex) {
```
❓ 에러를 `ex`로 표기하신 이유가 있나요?
`exception`, `e`로 많이 표현하는 것 같아서 여쭤봅니다!

hunnypooh replied:
e로 하니까 에러의 e같기도 해서 덜 헷갈릴려고 ex로 했습니다ㅎㅎ

kimyu0218 commented on .github/pull_request_template.md :
```diff
@@ -0,0 +1,19 @@
+<!-- 🔥 다음 양식으로 제목을 작성해주세요 : 한 일의 type[#issue number]: 작업 내용 -->
+<!-- ex) feat[#133]: canvas 구현~ -->
+<!-- "여기에 작성하세요" 는 지우고 작성하세요 🙏🏻 -->
+
+## 작업 개요
+<!-- 작업에 대한 설명을 간단하게 작성해주세요. -->
+여기에 작성하세요
+
+## 작업 사항
+<!-- 작업에 대한 설명을 코드와 관련하여 남겨주세요. -->
+여기에 작성하세요
+
+## 고민한 점들(필수 X)
+<!-- 작업을 진행하면서 고민했던 점들을 추가해주세요 -->
+여기에 작성하세요
+
+## 스크린샷(필수 X)
+<!-- 작업을 파악하는 데 도움이 되는 스크린샷을 추가해주세요 -->
+여기에 작성하세요
```
👍

kimyu0218 commented on src/main/java/com/brick/demo/config/WebSecurityConfig.java:
```diff
@@ -37,6 +37,8 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
             .requestMatchers("/auth/signin").permitAll()
             .requestMatchers("/auth/users/duplicate-email").permitAll()
+            .requestMatchers("/swagger-ui/**").permitAll()
+            .requestMatchers("/v3/api-docs/**").permitAll()
```
❓ `/v3/api-docs/**`는 뭔가요?

hunnypooh replied:
현재 Springdoc OpenAPI가 OpenAPI 3.0 버전에 맞게 자동으로 생성한 API 문서를 이용하는데, /v3/api-docs 이 경로를 통해 생성된 API 문서를 JSON 형식으로 변환해서 가져오고 있습니다. 기본 경로는 /v3/api-docs 으로 설정되고, 커스텀 할 수도 있다고 합니다ㅎㅎ

kimyu0218 commented on src/main/java/com/brick/demo/auth/service/AuthService.java:
```diff
@@ -70,9 +70,10 @@ public Optional<UserResponseDto> getAccountDetail(String authorizationHeader) {
-    boolean isDuplicate = accountManager.getAccountByEmail(dto.getEmail()) != null;
-    DuplicateEmailResponseDto response = new DuplicateEmailResponseDto(isDuplicate);
-    return response;
+    throw new CustomException(ErrorDetails.E001);
+//    boolean isDuplicate = accountManager.getAccountByEmail(dto.getEmail()) != null;
+//    DuplicateEmailResponseDto response = new DuplicateEmailResponseDto(isDuplicate);
+//    return response;
```
❓ 다 주석 처리하면 해당 코드는 어떻게 동작하는 건가요? 존재의 이유가 뭔가요?

hunnypooh replied:
깜박하고 못 지웠네요. 다시 지우고 다시 커밋 올렸습니다 ㅎㅎ

kimyu0218 reviewed (commented):
수고하셨습니다!!
몇가지 의문점이 있어서 질문 남겼습니다 :)

hunnypooh committed:
refactor: 불필요한 주석 삭제 [#3]

hunnypooh committed:
refactor: exclude url 중복 제거 [#3]

WebSecurityConfig 의 SecurityFilterChain 안의 authorizeHttpRequests, csrf 를 jwt filter chain에서 전달하는 exclude url 과 겹치는 부분은 exclude url 로 생성하도록 변경

hunnypooh committed:
fix: signin 예외 수정 [#3]

잘못된 이메일, 비밀 번호 입력했을때 보내는 예외 추가.

hunnypooh committed:
fix: duplicate-email 응답 수정 [#3]

hunnypooh committed:
feat: cors 설정 추가 [#3]
